migrations/src/main/scala/za/co/absa/enceladus/migrations/framework/migration/CollectionMigration.scala

```diff
@@ -0,0 +1,293 @@
+import za.co.absa.enceladus.migrations.framework.MigrationUtils
+import za.co.absa.enceladus.migrations.framework.dao.DocumentDb
+
+import scala.collection.mutable.ListBuffer
+
+/**
+  * A CollectionMigration represents an entity that provides ability to add, rename and remove collections
+  *
+  * In order to create a collection migration you need to extend from this trait and provide the requested
+  * collection changes:
+  *
+  * {{{
+  *   object MigrationTo1 extends MigrationBase with CollectionMigration {
+  *
+  *     addCollection("collection1_name")
+  *     addCollection("collection2_name")
+  *     addCollection("collection3_name")
+  *
+  *     removeCollection("collection4_name")
+  *     removeCollection("collection5_name")
+  *     removeCollection("collection6_name")
+  *
+  *     createIndex("collection1_name", "foo1" :: "bar1" :: Nil)
+  *     dropIndex("collection2_name", "foo2" :: "bar2" :: Nil)
+  *   }
+  * }}}
+  */
+trait CollectionMigration extends Migration {
+
+  type Index = (String, Seq[String])
+
+  private val log: Logger = LogManager.getLogger("CollectionMigration")
+
+  /**
+    * This method is used by derived classes to add new collection as a step of migration process.
+    *
+    * @param collectionName A collection name to be added
+    */
+  def createCollection(collectionName: String): Unit = {
+    if (collectionsToCreate.contains(collectionName)) {
+      throw new IllegalArgumentException(s"Collection '$collectionName' is already added.")
+    }
+    if (collectionsToDrop.contains(collectionName)) {
+      throw new IllegalArgumentException(s"Cannot both add and remove '$collectionName' as a migration step.")
+    }
+    collectionsToCreate += collectionName
+  }
+
+  /**
+    * This method is used by derived classes to remove a collection as a step of migration process.
+    *
+    * @param collectionName A collection name to be added
+    */
+  def dropCollection(collectionName: String): Unit = {
+    if (collectionsToDrop.contains(collectionName)) {
+      throw new IllegalArgumentException(s"Collection '$collectionName' is already in the removal list.")
+    }
+    if (collectionsToCreate.contains(collectionName)) {
+      throw new IllegalArgumentException(s"Cannot both add and remove '$collectionName' as a migration step.")
+    }
+    collectionsToDrop += collectionName
+  }
+
+  /**
+    * This method is used by derived classes to rename a collection as a step of migration process.
+    *
+    * @param oldName A collection to be renamed
+    * @param newName A new name for the collection
+    */
+  def renameCollection(oldName: String, newName: String): Unit = {
+    if (collectionsToDrop.contains(oldName)) {
+      throw new IllegalArgumentException(s"Collection '$oldName' is in the removal list.")
+    }
+    if (collectionsToCreate.contains(oldName)) {
+      throw new IllegalArgumentException(s"Collection '$oldName' is in the list of new collections. Cannot rename it.")
```

Comment:
Now I'm thinking if we need to check this here at all since validation does more sophisticated things... Better to check more and fail soon. So will add checks for the $newName as well.